docs/guides/kernel.md

```diff
@@ -1,82 +1,110 @@
 # Kernel
 
-This page explains how to compile or install linux-mbp-arch kernel and headers on non arch based distros. If you you are on an arch based distro, refer to [this section](#arch-based-systems-pacman). You may want to put this files into a package, so that you can manage the kernel files with your package manager. There is [some guidance](#packaging-the-kernel) about this, but the process is differs between package managers. If you have issues, make sure you are running the commands here with `bash`.
+This page explains how to compile a Linux kernel with patches for T2 hardware support and with apple-bce + apple-ibridge included (so no need to follow the [DKMS guide](https://wiki.t2linux.org/guides/dkms/) for this kernel). If you have issues, make sure you are running the commands here with `bash`.
 
-Before begining, set the directory that you want to install to.
+If your distro is not one of the distros with documentation on this Wiki, you may not need to compile a kernel yourself to get support for T2 hardware. Debian based systems can use the same kernel as described in the Ubuntu section, Arch based systems can use the same kernel as described in the Arch section, etc.
 
-If you want to install the kernel with a package manager:
+## Requirements
 
-```sh
-pkgdir=$PWD/pkg
-mkdir pkg
+- You will need some packages to build the kernel:
+
+    - Arch based systems: `sudo pacman --needed -S bc kmod libelf pahole cpio perl tar xz git`
+    - Debian based systems: `sudo apt install build-essential libncurses-dev libssl-dev flex bison libelf-dev bc dwarves openssl`
+    - For other distros you will need the equivalent of these, but if you miss something you'll most likely get an error saying what's missing, and you can then install it and re-run `make` to continue where you left off.
+
+- You will need about 20GB of disk space to compile the kernel. If you have a large amount of ram, you could use tmpfs to store build files in ram.
+
+## Building kernel
+
+### Getting kernel source and applying patches
+
+```bash
+mkdir build && cd build
+git clone --depth=1 https://github.com/Redecorating/mbp-16.1-linux-wifi patches
+source patches/PKGBUILD
+wget https://www.kernel.org/pub/linux/kernel/v${pkgver//.*}.x/linux-${pkgver}.tar.xz
+tar xf $_srcname.tar.xz
+cd $_srcname
+
+git clone --depth=1 https://github.com/t2linux/apple-bce-drv drivers/staging/apple-bce
+git clone --depth=1 https://github.com/t2linux/apple-ib-drv drivers/staging/apple-ibridge
+
+for patch in ../patches/*.patch; do
+    patch -Np1 < $patch
+done
 ```
 
-If you want to install to your current system directly (the kernel will not be managed by your package manager):
+### Setting kernel configuration
 
-```sh
-pkgdir=/
+```bash
+zcat /proc/config.gz > .config
```

Comment:
Ubuntu is not a rolling release distro. So, it still uses kernel config from way back 5.4. Thus for Ubuntu you may get the config by :-
```bash
curl -L https://github.com/AdityaGarg8/T2-Ubuntu-Kernel/raw/Mainline/templates/default-config > .config
sed -i 's/CONFIG_VERSION_SIGNATURE=.*/CONFIG_VERSION_SIGNATURE=""/g' .config
sed -i 's/CONFIG_SYSTEM_TRUSTED_KEYS=.*/CONFIG_SYSTEM_TRUSTED_KEYS=""/g' .config
sed -i 's/CONFIG_SYSTEM_REVOCATION_KEYS=.*/CONFIG_SYSTEM_REVOCATION_KEYS=""/g' .config
sed -i 's/CONFIG_DEBUG_INFO=y/# CONFIG_DEBUG_INFO is not set/g' .config
make olddefconfig
scripts/config --module apple-ibridge
scripts/config --module apple-bce
```

Reply:
it works fine when I grab the config from your kernel, but i dunno about older kernels and other distros

many distros have different configs so I'm not sure if it would work to have a long list of links to configs for recent kernels for a bunch of distros

If we use the arch config then that won't include, for example, the modules needed for anbox, which might confuse people on distros that normally support it (e.g. ubuntu). Maybe something like this could merge them, depending on how .config is parsed (i'll have to look into that when i get time)

```sh
cp ../patches/config .config
zcat /proc/config.gz >> .config
```

Does it fail to build if debian/canonical-certs.pem isn't there unless you remove CONFIG_SYSTEM_TRUSTED_KEYS? Instead of sed we could probably use scripts/config with --set-str and -u

Reply:
> it works fine when I grab the config from your kernel, but i dunno about older kernels and other distros
My kernel obviously will have the modified config so you won't face issues. But, some people prefer compiling on PCs having Linux support natively. They shall face issues then.
> 
> many distros have different configs so I'm not sure if it would work to have a long list of links to configs for recent kernels for a bunch of distros
> 
> If we use the arch config then that won't include, for example, the modules needed for anbox, which might confuse people on distros that normally support it (e.g. ubuntu). Maybe something like this could merge them, depending on how .config is parsed (i'll have to look into that when i get time)
> 
> ```shell
> cp ../patches/config .config
> zcat /proc/config.gz >> .config
> ```
> 
> Does it fail to build if debian/canonical-certs.pem isn't there unless you remove CONFIG_SYSTEM_TRUSTED_KEYS? Instead of sed we could probably use scripts/config with --set-str and -u
You may test it out if you get time.

Reply:
Do we have examples of using older config and then make olddefconfig breaking stuff?

Reply:
I don't have one. But old configs do cause non compilation of drivers that were upstreamed later, causing performance issues or lack of features. Eg:- I was unable to use exfat for a long time due to old kernel config.